### ISSUE TITLE
Kops GUI

### ISSUE THREAD

**Issue by yissachar** [kubernetes/kops#1159]
Step-by-step guides are nice but not everyone reads them. A GUI that manually walked the user through the steps required for cluster bring-up would ensure that all required steps were taken, and might be more user-friendly for kops beginners.

To start with I'm envisioning a very basic wizard with a few screens for choosing the various kops options (e.g. master size, instance size, zones), that just shells out to the kops CLI.

It's possible we would then want to expand this to cover all aspects of CLI, so viewing and editing the details of an existing kops-managed cluster as well.

**Comment by chrislovecnm** [comment]
Once we move kops to api machinery that is interesting.  All we need is a UX person...

**Comment by yissachar** [comment]
Can you clarify what "api machinery" is and how it relates to this?

**Comment by kris-nova** [comment]
@yissachar 

I have been thinking about this for a long time as well. Although I am not sure if we actually want a `gui` here? Can you give some clarity on **exactly** what you have in mind? Are we talking about a desktop appication? A proper backend `*curses` implementation? Or even just a simple web application that the user can run in a browser that stores it's source in the repository?

Basically there seems to be a few approaches to this in my mind, and just wondering if you could give more details. Also you mentioned the step-by-step guides, is this a wizard used for learning only? Or do we actually plan on interfacing with the program here and offering this new interface as a supported way to interact with kops?

If we are talking about a desktop application here you admittedly have my full attention..

Regarding the api machinery comment,

Basically kops offers a library API mirrored from k8s itself, we are in the works of implementing some back end machinery that will allow us to swap out API versions like our big brother k8s. To be honest I don't know why that applies here. Maybe something to do with exposing certain elements of kops to the user in the new GUI?

**Comment by yissachar** [comment]
I was thinking either a desktop app (possibly Electron based), or a browser based app.  I would prefer the desktop app approach, assuming that it doesn't increase the size of kops too much (if we go with Electron we are basically bundling a browser in there).

The ambitious end-goal is to have an alternative way to interact with kops, through a GUI as opposed to do the current CLI. So this app would be able to do everything that the CLI currently does: create a cluster, view cluster settings, edit cluster settings, upgrade Kubernetes version, delete cluster, etc. The wizard would be for creating a cluster.

Possible advantages of a GUI approach:

1.  We can display docs inline where the user choose an option
2. Settings would be more discoverable than editing a YAML file (e.g. networking options would be a dropdown list)
3. Easier to get initial cluster setup right, through the wizard

Obviously the main disadvantage will be the effort involved in building and maintaining this GUI.

**Comment by chrislovecnm** [comment]
Your last sentence is key :) Before we venture this direction, we would need core contributors that can maintain this functionality long term.

Great idea, but the question is who.

Also this is not a small task. Per our description this would need to be production grade UI.

**Comment by kris-nova** [comment]
I think this is the right time to bring this up, even if we don't start on it now, making critical engineering choices around the idea for the future will be important.

Particularly things like documentation, and order of operations.. We spend a lot of time / effort on those areas right now, and having a linear process that we can control might be helpful.

So it may not be a wash, but we would be exchanging dev time on issues we are already work on, for a more viable long term solution. 

I think it's good to bring up. We are getting more and more attention (which means more and more potential contributors), and this could help the community quite a bit.
